server/src/utils/media.ts

```diff
@@ -979,12 +988,21 @@ export class RkmppHwDecodeConfig extends RkmppSwDecodeConfig {
   getFilterOptions(videoStream: VideoStreamInfo) {
     if (this.shouldToneMap(videoStream)) {
       const { primaries, transfer, matrix } = this.getColors();
-      return [
-        `scale_rkrga=${this.getScaling(videoStream)}:format=p010:afbc=1`,
-        'hwmap=derive_device=opencl:mode=read',
-        `tonemap_opencl=format=nv12:r=pc:p=${primaries}:t=${transfer}:m=${matrix}:tonemap=${this.config.tonemap}:desat=0:tonemap_mode=lum:peak=100`,
-        'hwmap=derive_device=rkmpp:mode=write:reverse=1',
-        'format=drm_prime',
+      if (this.hasMaliOpenCL) {
+        return [ // use RKMPP for scaling, OpenCL for tone mapping
+          `scale_rkrga=${this.getScaling(videoStream)}:format=p010:afbc=1`,
+          'hwmap=derive_device=opencl:mode=read',
+          `tonemap_opencl=format=nv12:r=pc:p=${primaries}:t=${transfer}:m=${matrix}:tonemap=${this.config.tonemap}:desat=0:tonemap_mode=lum:peak=100`,
+          'hwmap=derive_device=rkmpp:mode=write:reverse=1',
+          'format=drm_prime',
+        ];
+      }
+      return [ // use RKMPP for scaling, CPU for tone mapping
+        `scale_rkrga=${this.getScaling(videoStream)}:format=nv12`,
+        'hwdownload',
+        'format=nv12',
```

Comment:
You can't set the format to nv12 before tone-mapping - it will alter the colors before tone-mapping has a chance to map them to SDR. Does it not work without setting a format here? If not, use yuv420p10le instead to keep it 10-bit.